### ISSUE TITLE
Native image build failure: undefined reference to 'dlsym'

### ISSUE THREAD

**Issue by dsyer** [oracle/graal#1091]
```
[target/demo:325]    classlist:   2,255.09 ms
[target/demo:325]        (cap):     783.05 ms
[target/demo:325]        setup:   1,614.00 ms
Warning: class initialization of class reactor.netty.http.client.HttpClientSecure failed with exception java.lang.NoClassDefFoundError: io/netty/handler/codec/http2/Http2SecurityUtil. This class will be initialized at run time because either option --report-unsupported-elements-at-runtime or option --allow-incomplete-classpath is used for image building. Use the option --delay-class-initialization-to-runtime=reactor.netty.http.client.HttpClientSecure to explicitly request delayed initialization of this class.
Warning: class initialization of class org.springframework.http.codec.xml.XmlEventDecoder$AaltoDataBufferToXmlEvent failed with exception java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: com/fasterxml/aalto/stax/InputFactoryImpl. This class will be initialized at run time because either option --report-unsupported-elements-at-runtime or option --allow-incomplete-classpath is used for image building. Use the option --delay-class-initialization-to-runtime=org.springframework.http.codec.xml.XmlEventDecoder$AaltoDataBufferToXmlEvent to explicitly request delayed initialization of this class.
[target/demo:325]   (typeflow):   8,321.13 ms
[target/demo:325]    (objects):   6,993.90 ms
[target/demo:325]   (features):     216.48 ms
[target/demo:325]     analysis:  15,934.38 ms
[target/demo:325]     universe:     453.71 ms
[target/demo:325]      (parse):   1,488.93 ms
[target/demo:325]     (inline):   1,919.93 ms
[target/demo:325]    (compile):   9,589.53 ms
[target/demo:325]      compile:  13,789.60 ms
[target/demo:325]        image:   1,245.44 ms
[target/demo:325]        write:     172.54 ms
Fatal error: java.lang.RuntimeException: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1

Running command: cc -v -o /home/dsyer/dev/scratch/graal/graal-issues/rest/target/demo -z noexecstack -Wl,--defsym -Wl,main=IsolateEnterStub__JavaMainWrapper__run__5087f5482cc9a6abc971913ece43acb471d2631b__a61fe6c26e84dd4037e4629852b5488bfcc16e7e -L/tmp/SVM-4127825426168344561 -L/home/dsyer/.sdkman/candidates/java/1.0.0-rc-14-grl/jre/lib -L/home/dsyer/.sdkman/candidates/java/1.0.0-rc-14-grl/jre/lib/svm/clibraries/linux-amd64 /tmp/SVM-4127825426168344561/target/demo.o -lffi -ldl -lpthread -lpthread -lz -lpthread -ldl -lstrictmath -ljvm -llibchelper -lcrypt -lrt -lrt -lrt -lrt -lrt

Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/7/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 7.3.0-27ubuntu1~18.04' --with-bugurl=file:///usr/share/doc/gcc-7/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-7 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 7.3.0 (Ubuntu 7.3.0-27ubuntu1~18.04) 
COMPILER_PATH=/usr/lib/gcc/x86_64-linux-gnu/7/:/usr/lib/gcc/x86_64-linux-gnu/7/:/usr/lib/gcc/x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/7/:/usr/lib/gcc/x86_64-linux-gnu/
LIBRARY_PATH=/usr/lib/gcc/x86_64-linux-gnu/7/:/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/:/usr/lib/gcc/x86_64-linux-gnu/7/../../../../lib/:/lib/x86_64-linux-gnu/:/lib/../lib/:/usr/lib/x86_64-linux-gnu/:/usr/lib/../lib/:/usr/lib/gcc/x86_64-linux-gnu/7/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-v' '-o' '/home/dsyer/dev/scratch/graal/graal-issues/rest/target/demo' '-z' 'noexecstack' '-L/tmp/SVM-4127825426168344561' '-L/home/dsyer/.sdkman/candidates/java/1.0.0-rc-14-grl/jre/lib' '-L/home/dsyer/.sdkman/candidates/java/1.0.0-rc-14-grl/jre/lib/svm/clibraries/linux-amd64' '-mtune=generic' '-march=x86-64'
 /usr/lib/gcc/x86_64-linux-gnu/7/collect2 -plugin /usr/lib/gcc/x86_64-linux-gnu/7/liblto_plugin.so -plugin-opt=/usr/lib/gcc/x86_64-linux-gnu/7/lto-wrapper -plugin-opt=-fresolution=/tmp/ccVja4sR.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --sysroot=/ --build-id --eh-frame-hdr -m elf_x86_64 --hash-style=gnu --as-needed -dynamic-linker /lib64/ld-linux-x86-64.so.2 -pie -z now -z relro -o /home/dsyer/dev/scratch/graal/graal-issues/rest/target/demo -z noexecstack /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/Scrt1.o /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/7/crtbeginS.o -L/tmp/SVM-4127825426168344561 -L/home/dsyer/.sdkman/candidates/java/1.0.0-rc-14-grl/jre/lib -L/home/dsyer/.sdkman/candidates/java/1.0.0-rc-14-grl/jre/lib/svm/clibraries/linux-amd64 -L/usr/lib/gcc/x86_64-linux-gnu/7 -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/7/../../.. --defsym main=IsolateEnterStub__JavaMainWrapper__run__5087f5482cc9a6abc971913ece43acb471d2631b__a61fe6c26e84dd4037e4629852b5488bfcc16e7e /tmp/SVM-4127825426168344561/target/demo.o -lffi -ldl -lpthread -lpthread -lz -lpthread -ldl -lstrictmath -ljvm -llibchelper -lcrypt -lrt -lrt -lrt -lrt -lrt -lgcc --push-state --as-needed -lgcc_s --pop-state -lc -lgcc --push-state --as-needed -lgcc_s --pop-state /usr/lib/gcc/x86_64-linux-gnu/7/crtendS.o /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/crtn.o
/home/dsyer/.sdkman/candidates/java/1.0.0-rc-14-grl/jre/lib/svm/clibraries/linux-amd64/libjvm.a(JvmFuncs.o): In function `JVM_FindLibraryEntry':
/b/b/e/main/substratevm/src/com.oracle.svm.native.jvm.posix/src/JvmFuncs.c:112: undefined reference to `dlsym'
collect2: error: ld returned 1 exit status

	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
	at java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1005)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:433)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:284)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:406)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.executeCompilation(NativeImageBuildServer.java:390)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.lambda$processCommand$8(NativeImageBuildServer.java:327)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.withJVMContext(NativeImageBuildServer.java:408)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.processCommand(NativeImageBuildServer.java:324)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.processRequest(NativeImageBuildServer.java:268)
	at com.oracle.svm.hosted.server.NativeImageBuildServer.lambda$serve$7(NativeImageBuildServer.java:228)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1

```

Sample to reproduce (subdir called "rest"): https://github.com/sdeleuze/graal-issues/pull/4

See #1086 - looks like the same error but I didn't use `--static`.

**Comment by olpaw** [comment]
Hmm, for some reason i cannot reproduce the issue with RC14
```
[master=] ~/OLabs/git/graal-issues/rest> ./build.sh 
[INFO] Scanning for projects...
[INFO] 
[INFO] --------------------------< com.sample:rest >---------------------------
[INFO] Building rest 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ rest ---
[INFO] Deleting /home/pwoegere/OLabs/git/graal-issues/rest/target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ rest ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/pwoegere/OLabs/git/graal-issues/rest/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ rest ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to /home/pwoegere/OLabs/git/graal-issues/rest/target/classes
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ rest ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/pwoegere/OLabs/git/graal-issues/rest/src/test/resources
[INFO] 
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ rest ---
[INFO] No sources to compile
[INFO] 
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ rest ---
[INFO] No tests to run.
[INFO] 
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ rest ---
[INFO] Building jar: /home/pwoegere/OLabs/git/graal-issues/rest/target/rest-1.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-shade-plugin:3.2.1:shade (default) @ rest ---
[INFO] Including org.springframework:spring-webflux:jar:5.1.4.RELEASE in the shaded jar.
[INFO] Including io.projectreactor:reactor-core:jar:3.2.5.RELEASE in the shaded jar.
[INFO] Including org.reactivestreams:reactive-streams:jar:1.0.2 in the shaded jar.
[INFO] Including org.springframework:spring-beans:jar:5.1.4.RELEASE in the shaded jar.
[INFO] Including org.springframework:spring-core:jar:5.1.4.RELEASE in the shaded jar.
[INFO] Including org.springframework:spring-jcl:jar:5.1.4.RELEASE in the shaded jar.
[INFO] Including org.springframework:spring-web:jar:5.1.4.RELEASE in the shaded jar.
[INFO] Including io.projectreactor.netty:reactor-netty:jar:0.8.3.RELEASE in the shaded jar.
[INFO] Including io.netty:netty-codec-http:jar:4.1.31.Final in the shaded jar.
[INFO] Including io.netty:netty-codec:jar:4.1.31.Final in the shaded jar.
[INFO] Including io.netty:netty-handler:jar:4.1.31.Final in the shaded jar.
[INFO] Including io.netty:netty-buffer:jar:4.1.31.Final in the shaded jar.
[INFO] Including io.netty:netty-common:jar:4.1.31.Final in the shaded jar.
[INFO] Including io.netty:netty-transport:jar:4.1.31.Final in the shaded jar.
[INFO] Including io.netty:netty-resolver:jar:4.1.31.Final in the shaded jar.
[INFO] Including org.slf4j:slf4j-jdk14:jar:1.7.25 in the shaded jar.
[INFO] Including org.slf4j:slf4j-api:jar:1.7.25 in the shaded jar.
[INFO] Replacing original artifact with shaded artifact.
[INFO] Replacing /home/pwoegere/OLabs/git/graal-issues/rest/target/rest-1.0-SNAPSHOT.jar with /home/pwoegere/OLabs/git/graal-issues/rest/target/rest-1.0-SNAPSHOT-shaded.jar
[INFO] Dependency-reduced POM written at: /home/pwoegere/OLabs/git/graal-issues/rest/dependency-reduced-pom.xml
[INFO] 
[INFO] --- maven-install-plugin:2.4:install (default-install) @ rest ---
[INFO] Installing /home/pwoegere/OLabs/git/graal-issues/rest/target/rest-1.0-SNAPSHOT.jar to /home/pwoegere/.m2/repository/com/sample/rest/1.0-SNAPSHOT/rest-1.0-SNAPSHOT.jar
[INFO] Installing /home/pwoegere/OLabs/git/graal-issues/rest/dependency-reduced-pom.xml to /home/pwoegere/.m2/repository/com/sample/rest/1.0-SNAPSHOT/rest-1.0-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 2.723 s
[INFO] Finished at: 2019-03-21T14:41:32+01:00
[INFO] ------------------------------------------------------------------------
Build on Server(pid: 12454, port: 42051)*
[target/demo:12454]    classlist:   2,765.90 ms
[target/demo:12454]        (cap):     673.06 ms
[target/demo:12454]        setup:   1,697.44 ms
Warning: class initialization of class reactor.netty.http.client.HttpClientSecure failed with exception java.lang.NoClassDefFoundError: io/netty/handler/codec/http2/Http2SecurityUtil. This class will be initialized at run time because either option --report-unsupported-elements-at-runtime or option --allow-incomplete-classpath is used for image building. Use the option --delay-class-initialization-to-runtime=reactor.netty.http.client.HttpClientSecure to explicitly request delayed initialization of this class.
Warning: class initialization of class org.springframework.http.codec.xml.XmlEventDecoder$AaltoDataBufferToXmlEvent failed with exception java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: com/fasterxml/aalto/stax/InputFactoryImpl. This class will be initialized at run time because either option --report-unsupported-elements-at-runtime or option --allow-incomplete-classpath is used for image building. Use the option --delay-class-initialization-to-runtime=org.springframework.http.codec.xml.XmlEventDecoder$AaltoDataBufferToXmlEvent to explicitly request delayed initialization of this class.
[target/demo:12454]   (typeflow):   9,301.77 ms
[target/demo:12454]    (objects):   7,138.28 ms
[target/demo:12454]   (features):     231.61 ms
[target/demo:12454]     analysis:  17,116.77 ms
[target/demo:12454]     universe:     516.35 ms
[target/demo:12454]      (parse):   1,996.33 ms
[target/demo:12454]     (inline):   1,660.94 ms
[target/demo:12454]    (compile):  11,394.38 ms
[target/demo:12454]      compile:  16,007.56 ms
[target/demo:12454]        image:   1,669.90 ms
[target/demo:12454]        write:     244.44 ms
[target/demo:12454]      [total]:  40,182.99 ms
In regular JVM
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="http://www.google.com/">here</A>.
</BODY></HTML>


As GraalVM native image
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="http://www.google.com/">here</A>.
</BODY></HTML>
```
LGTM. I tested with
```
[master=] ~/OLabs/git/graal-issues/rest> native-image --version
GraalVM Version 1.0.0-rc14
[master=] ~/OLabs/git/graal-issues/rest> which native-image 
~/OLabs/graalvm-ce-1.0.0-rc14/bin/native-image
```

Anyway, fixing #1086 should also solve your problem here.

**Comment by olpaw** [comment]
Fixed on master (5dbe774178c59e34f886fe884dc12cb7f0f6110b)
